Assets/VIRTUOSO/Scripts/HapticSDK/Haptic Devices/UnityDevice.cs

```diff
@@ -0,0 +1,131 @@
+﻿using System.Collections;
+using System.Collections.Generic;
+using UnityEngine;
+using VRTK;
+
+#if UNITY_2017_2_OR_NEWER
+using UnityEngine.XR;
+#else
+    using UnityEngine.VR;
+    using XRNode = UnityEngine.VR.VRNode;
+#endif
+
+namespace CharlesRiverAnalytics.Virtuoso.Haptic
+{
+    /// <summary>
+    /// Haptic device for Unity XR input devices. To use, first make sure a scene has a Haptic Manager.
+    /// Then, set up two objects with UnityDevice scripts (one for each controller). While not required,
+    /// attaching these to the SDK Unity Controller Tracker objects is preferred.
+    /// 
+    /// Author: Dan Duggan (dduggan@cra.com) Updated: September 2019
+    /// </summary>
+    [HapticSystem("Unity", "Left Controller", "BodyCoordinates/LeftHand", typeof(SDK_UnityController), false)]
+    [HapticSystem("Unity", "Right Controller", "BodyCoordinates/RightHand", typeof(SDK_UnityController), true)]
+    public class UnityDevice : HapticDevice
+    {
+#if UNITY_2018_3_OR_NEWER
+        #region Protected Variables
+        // Indicates which hand this object corresponds to
+        protected XRNode nodeType;
+        
+        // Holds a reference to the Unity XR device for haptic feedback
+        protected InputDevice hapticDevice;
+
+        // By default, this is set to zero
+        protected uint hapticChannel = 0;
+        #endregion
+
+        #region Haptic Device Behavior
+        protected override void CancelHaptics()
+        {
+#if UNITY_2019_1_OR_NEWER
+            if(hapticDevice.isValid)
+#elif UNITY_2018_3_OR_NEWER
+            if(hapticDevice.IsValid)
+#endif
+            {
+                hapticDevice.StopHaptics();
+            }
+            else
+            {
+                Debug.LogWarningFormat("Unity XR Haptic Device {0} failed to cancel: device is invalid", nodeType.ToString());
+            }
+        }
+
+        protected override void StartHaptics(HumanBodyBones bodyPart, BodyCoordinateHit hitLocation, float intensity)
+        {
+#if UNITY_2019_1_OR_NEWER
```

Comment:
I'd create a helper function that returns a boolean - something like "IsHapticDeviceValid" and do the platform checks within that.

Reply:
Good call, the defines were getting out of hand